src/cloud.c

```diff
@@ -56,60 +56,83 @@
 #define AMQP_CMD_DATA_PUBLISH "data.publish"
 #define AMQP_CMD_DEVICE_UNREGISTER "device.unregister"
 
-struct cloud_callbacks {
+/* Hashmap to user callback from southbound traffic */
+static struct l_hashmap *map_event_to_cb;
+
+/* Union with all cloud callbacks */
+union cloud_cb_t {
 	cloud_downstream_cb_t update_cb;
-	cloud_downstream_cb_t request_cb;
 	cloud_device_removed_cb_t removed_cb;
 };
 
-static struct cloud_callbacks cloud_cbs;
+/* Struct that has a callback #cb that handle event in #event_id */
+struct event_handler {
+	void *cb;
+	enum {
+		UPDATE_EVT,
+		REQUEST_EVT,
+		REMOVED_EVT
+	} event_id;
+};
+
+static const char *get_id_from_json_obj(json_object *jso)
+{
+	json_object *jobjkey;
+
+	if (!json_object_object_get_ex(jso, "id", &jobjkey))
+		return false;
+
+	if (json_object_get_type(jobjkey) != json_type_string)
+		return false;
+
+	return json_object_get_string(jobjkey);
+}
 
 static bool cloud_receive_message(const char *exchange,
 				  const char *routing_key,
 				  const char *body, void *user_data)
 {
 	json_object *jso;
-	json_object *id_json;
 	struct l_queue *list;
 	const char *id;
+	struct event_handler *handler;
+	union cloud_cb_t cb_handler;
 
 	if (strcmp(AMQP_EXCHANGE_FOG, exchange) != 0)
 		return true;
 
-	jso = json_tokener_parse(body);
-	if (!jso)
+	handler = l_hashmap_lookup(map_event_to_cb, routing_key);
+	if (!handler->cb)
 		return false;
 
-	if (!json_object_object_get_ex(jso, "id", &id_json))
-		return false;
-
-	if (json_object_get_type(id_json) != json_type_string)
+	jso = json_tokener_parse(body);
+	if (!jso)
 		return false;
 
-	id = json_object_get_string(id_json);
-
-	if (cloud_cbs.update_cb != NULL &&
-	    strcmp(AMQP_EVENT_DATA_UPDATE, routing_key) == 0) {
+	switch (handler->event_id) {
+	case UPDATE_EVT:
+		cb_handler.update_cb = handler->cb;
+		id = get_id_from_json_obj(jso);
 		list = parser_update_to_list(jso);
 		if (list == NULL) {
 			hal_log_error("Error on parse json object");
 			json_object_put(jso);
 			return false;
 		}
 
-		cloud_cbs.update_cb(id, list, user_data);
+		cb_handler.update_cb(id, list, user_data);
 		l_queue_destroy(list, l_free);
-	}
-
-	if (cloud_cbs.request_cb != NULL &&
-	    strcmp(AMQP_EVENT_DATA_REQUEST, routing_key) == 0) {
+		break;
+	case REQUEST_EVT:
 		return true;
```

Comment:
I would change the `return true;` for a `break;`.

Reply:
Use break because of the json_object_put(jso); on line 134

Reply:
Done